plugin/pkg/admission/initialization/initialization.go

```diff
@@ -0,0 +1,173 @@
+}
+
+type initializer struct {
+	resources  map[schema.GroupResource]initializerOptions
+	authorizer authorizer.Authorizer
+}
+
+// NewAlwaysAdmit creates a new always admit admission handler
+func NewInitializer() admission.Interface {
+	return &initializer{
+		resources: map[schema.GroupResource]initializerOptions{
+		//schema.GroupResource{Resource: "pods"}: {Initializers: []string{"Test"}},
+		},
+	}
+}
+
+func (i *initializer) Validate() error {
+	if i.authorizer == nil {
+		return fmt.Errorf("requires authorizer")
+	}
+	return nil
+}
+
+func (i *initializer) SetAuthorizer(a authorizer.Authorizer) {
+	i.authorizer = a
+}
+
+func (i *initializer) Admit(a admission.Attributes) (err error) {
+	// TODO: sub-resource action should be denied until the object is initialized
+	if len(a.GetSubresource()) > 0 {
+		return nil
+	}
+
+	resource, ok := i.resources[a.GetResource().GroupResource()]
+	if !ok {
+		return nil
+	}
+
+	switch a.GetOperation() {
+	case admission.Create:
+		accessor, err := meta.Accessor(a.GetObject())
+		if err != nil {
+			return fmt.Errorf("initialized resources must be able to set initializers (%T): %v", a.GetObject(), err)
+		}
+		existing := accessor.GetInitializers()
+		// it must be possible for some users to bypass initialization - for now, check the initialize operation
+		if existing != nil {
+			if err := i.canInitialize(a); err != nil {
+				return err
+			}
+		}
+
+		// TODO: pull this from config
+		accessor.SetInitializers(copiedInitializers(resource.Initializers))
+
+	case admission.Update:
+		accessor, err := meta.Accessor(a.GetObject())
+		if err != nil {
+			return fmt.Errorf("initialized resources must be able to set initializers (%T): %v", a.GetObject(), err)
+		}
+		updated := accessor.GetInitializers()
+
+		existingAccessor, err := meta.Accessor(a.GetOldObject())
+		if err != nil {
+			return fmt.Errorf("initialized resources must be able to set initializers (%T): %v", a.GetOldObject(), err)
+		}
+		existing := existingAccessor.GetInitializers()
+
+		// post initialization, all changes are allowed
+		if existing == nil && updated == nil {
+			return nil
+		}
+		// you cannot re-initialize an initialized object
+		if existing == nil && updated != nil {
+			return errors.NewInvalid(a.GetKind().GroupKind(), a.GetName(), field.ErrorList{field.Invalid(field.NewPath("metadata", "initializers"), nil, "initializers may not be added to already initialized objects")})
+		}
+		// reject mutation of the object after initialization failure status is set
+		if existing.Result != nil && existing.Result.Status == metav1.StatusFailure {
+			return errors.NewForbidden(a.GetResource().GroupResource(), a.GetName(), fmt.Errorf("this object failed initialization and cannot be updated: %v", existing.Result.Message))
+		}
+		// reject any update where initializers is set, pending is nil, and the result isn't failure
+		if len(updated.Pending) == 0 && (updated.Result == nil || updated.Result.Status != metav1.StatusFailure) {
```

Comment:
I don't think we want to - I want to explicitly allow only three possible writes from initialization controllers:

1. remove / add initializers
2. mark initialization complete by removing the entire struct
3. mark initialization failed by setting status and clearing pending

I would prefer to be more restrictive.